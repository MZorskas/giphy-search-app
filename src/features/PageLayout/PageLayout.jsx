import React, { useState, useCallback, useEffect } from 'react';
import './PageLayout.scss';
import Header from 'features/Header';
import GalleryContainer from 'features/GalleryContainer';
import giphy from 'apis/giphy';
import { ReactComponent as Loader } from 'assets/ellipsis_loader.svg';
import ErrorMessage from 'components/ErrorMessage';

const PageLayout = () => {
  const [hasData, setHasData] = useState(false);
  const [data, setData] = useState([]);
  const [searchValue, setSearchValue] = useState();
  const [totalCount, setTotalCount] = useState(0);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState();

  const fetchData = useCallback(async (value, offset = '0') => {
    setIsLoading(true);
    setError('');
    setSearchValue(value);
    try {
      const response = await giphy.get('/search', {
        params: {
          q: value,
          offset: offset,
        },
      });
      setIsLoading(false);
      setHasData(true);
      setData((previousData) => previousData.concat(response.data.data));
      setTotalCount(response.data.pagination.total_count);
    } catch (e) {
      setIsLoading(false);
      setError(e.message);
    }
  }, []);

  useEffect(() => {
    setData([]);
    setHasData(false);
  }, [searchValue]);

  return (
    <main className="page-layout" data-testid="page-layout">
      <div className="page-layout__content-wrapper">
        <Header fetchData={fetchData} />
        {hasData && (
          <GalleryContainer
            title={searchValue}
            totalCount={totalCount}
            data={data}
            fetchData={fetchData}
            error={error}
            isLoading={isLoading}
          />
        )}
        {error && !isLoading && <ErrorMessage message={error} />}
        {isLoading && <Loader className="page-layout__loader" />}
      </div>
    </main>
  );
};

export default PageLayout;
