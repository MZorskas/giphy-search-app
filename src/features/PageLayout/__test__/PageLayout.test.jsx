import React from 'react';
import PageLayout from '../PageLayout';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

const mockfetchData = jest.fn();

const totalCount = 10;
const title = 'Title';
const errorMessage = `No GIFs found for ${title}`;
const altText = 'Test';
const data = [
  {
    images: { url: 'url1' },
    title: altText,
  },
  {
    images: { url: 'url2' },
    title: altText,
  },
  {
    images: { url: 'url3' },
    title: altText,
  },
];

jest.mock('Features/Header', () => (props) => (
  <mock-Header data-testId="mock-header" {...props} />
));

jest.mock('Features/GalleryContainer', () => (props) => (
  <mock-GalleryContainer data-testId="mock-gallery-container" {...props} />
));

describe('PageLayout component', () => {
  it('renders', () => {
    expect(screen.queryByTestId('page-layout')).not.toBeInTheDocument();
    render(<PageLayout />);
    expect(screen.queryByTestId('page-layout')).toBeInTheDocument();
  });
  //   it('renders without propTypes warning', () => {
  //     const spy = jest.spyOn(global.console, 'error');
  //     expect(screen.queryByTestId('gallery-container')).not.toBeInTheDocument();
  //     render(
  //       <GalleryContainer
  //         data={[]}
  //         fetchData={mockfetchData}
  //         title={title}
  //         totalCount={0}
  //       />
  //     );
  //     expect(screen.queryByTestId('gallery-container')).toBeInTheDocument();
  //     expect(spy).not.toHaveBeenCalled();
  //   });
  //   it('displays title and totalCount correctly', () => {
  //     expect(screen.queryByText(title)).not.toBeInTheDocument();
  //     expect(screen.queryByText(`${totalCount} GIFs`)).not.toBeInTheDocument();
  //     render(
  //       <GalleryContainer
  //         data={[]}
  //         fetchData={mockfetchData}
  //         title={title}
  //         totalCount={totalCount}
  //       />
  //     );
  //     expect(screen.queryByText(title)).toBeInTheDocument();
  //     expect(screen.queryByText(`${totalCount} GIFs`)).toBeInTheDocument();
  //   });
  //   it('renders ImageCard elements', () => {
  //     expect(screen.queryByTestId('mock-image-card')).not.toBeInTheDocument();
  //     render(
  //       <GalleryContainer
  //         data={data}
  //         fetchData={mockfetchData}
  //         title={title}
  //         totalCount={totalCount}
  //       />
  //     );
  //     expect(screen.getAllByTestId('mock-image-card')[0]).toHaveAttribute(
  //       'alt',
  //       data[0].title
  //     );
  //     expect(screen.getAllByTestId('mock-image-card').length).toBe(data.length);
  //   });
  //   it('displays ErrorMessage if totalCount is 0', () => {
  //     expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();
  //     const { rerender } = render(
  //       <GalleryContainer
  //         data={[]}
  //         fetchData={mockfetchData}
  //         title={title}
  //         totalCount={totalCount}
  //       />
  //     );
  //     expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();
  //     rerender(
  //       <GalleryContainer
  //         data={[]}
  //         fetchData={mockfetchData}
  //         title={title}
  //         totalCount={0}
  //       />
  //     );
  //     expect(screen.queryByText(errorMessage)).toBeInTheDocument();
  //   });
  //   it('displays ErrorMessage with correct title', () => {
  //     expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();
  //     const { rerender } = render(
  //       <GalleryContainer
  //         data={[]}
  //         fetchData={mockfetchData}
  //         title={title}
  //         totalCount={0}
  //       />
  //     );
  //     expect(
  //       screen.queryByText(`No GIFs found for ${title}`)
  //     ).toBeInTheDocument();
  //     rerender(
  //       <GalleryContainer
  //         data={[]}
  //         fetchData={mockfetchData}
  //         title="Test"
  //         totalCount={0}
  //       />
  //     );
  //     expect(
  //       screen.queryByText(`No GIFs found for ${title}`)
  //     ).not.toBeInTheDocument();
  //     // expect(screen.queryByText(`No GIFs found for Test`)).toBeInTheDocument();
  //     expect(screen.queryByText(`No GIFs found for Test`)).toBeInTheDocument();
  //     expect;
  //   });
  it('matches snapshot ', () => {
    const tree = renderer.create(<PageLayout />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  //   it('matches snapshot with data', () => {
  //     const tree = renderer
  //       .create(
  //         <GalleryContainer
  //           data={data}
  //           fetchData={mockfetchData}
  //           title="Test"
  //           totalCount={totalCount}
  //         />
  //       )
  //       .toJSON();
  //     expect(tree).toMatchSnapshot();
  //   });
  //   it('matches snapshot with ErrorMessage', () => {
  //     const tree = renderer
  //       .create(
  //         <GalleryContainer
  //           data={[]}
  //           fetchData={mockfetchData}
  //           title="Test"
  //           totalCount={0}
  //         />
  //       )
  //       .toJSON();
  //     expect(tree).toMatchSnapshot();
  //   });
});
