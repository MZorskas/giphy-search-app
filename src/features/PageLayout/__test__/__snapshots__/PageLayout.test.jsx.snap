// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageLayout component matches snapshot  1`] = `
<main
  className="page-layout"
  data-testid="page-layout"
>
  <div
    className="page-layout__content-wrapper"
  >
    <mock-Header
      data-testId="mock-header"
      fetchData={[Function]}
    />
  </div>
</main>
`;
